components/script/dom/readablestream.rs

```diff
@@ -0,0 +1,517 @@
+            ExternalUnderlyingSource::FetchResponse => vec![],
+        };
+        ExternalUnderlyingSourceController {
+            buffer: RefCell::new(buffer),
+            closed: RefCell::new(false),
+        }
+    }
+
+    /// Signal available bytes if the stream is currently readable.
+    #[allow(unsafe_code)]
+    fn maybe_signal_available_bytes(
+        &self,
+        cx: SafeJSContext,
+        stream: HandleObject,
+        available: usize,
+    ) {
+        if available == 0 {
+            return;
+        }
+        unsafe {
+            let mut readable = false;
+            if !ReadableStreamIsReadable(*cx, stream, &mut readable) {
+                return;
+            }
+            if readable {
+                ReadableStreamUpdateDataAvailableFromSource(*cx, stream, available as u32);
+            }
+        }
+    }
+
+    /// Close a currently readable js stream.
+    #[allow(unsafe_code)]
+    fn maybe_close_js_stream(&self, cx: SafeJSContext, stream: HandleObject) {
+        unsafe {
+            let mut readable = false;
+            if !ReadableStreamIsReadable(*cx, stream, &mut readable) {
+                return;
+            }
+            if readable {
+                ReadableStreamClose(*cx, stream);
+            }
+        }
+    }
+
+    fn close(&self, cx: SafeJSContext, stream: HandleObject) {
+        *self.closed.borrow_mut() = true;
+        self.maybe_close_js_stream(cx, stream);
+    }
+
+    fn enqueue_chunk(&self, cx: SafeJSContext, stream: HandleObject, mut chunk: Vec<u8>) {
+        let available = {
+            let mut buffer = self.buffer.borrow_mut();
+            buffer.append(&mut chunk);
+            buffer.len()
+        };
+        self.maybe_signal_available_bytes(cx, stream, available);
+    }
+
+    #[allow(unsafe_code)]
+    fn pull(&self, cx: SafeJSContext, stream: HandleObject, _desired_size: usize) {
+        // Note: for pull sources,
+        // this would be the time to ask for a chunk.
+
+        if { *self.closed.borrow() } {
+            return self.maybe_close_js_stream(cx, stream);
+        }
+
+        let available = {
+            let buffer = self.buffer.borrow();
+            buffer.len()
+        };
+
+        self.maybe_signal_available_bytes(cx, stream, available);
+    }
+
+    fn get_chunk_with_length(&self, length: usize) -> Vec<u8> {
+        let mut buffer = self.buffer.borrow_mut();
+        let buffer_len = buffer.len();
+        assert!(buffer_len >= length as usize);
+        buffer.split_off(buffer_len - length)
+    }
+
+    #[allow(unsafe_code)]
+    fn write_into_buffer(
+        &self,
+        _cx: SafeJSContext,
+        _stream: HandleObject,
+        buffer: *mut c_void,
+        length: usize,
+        bytes_written: *mut usize,
+    ) {
+        let chunk = self.get_chunk_with_length(length);
+
+        unsafe {
+            *bytes_written = chunk.len();
+            ptr::copy_nonoverlapping(chunk.as_ptr(), buffer as *mut u8, chunk.len());
```

Comment:
If buffer is a slice, we can use https://doc.rust-lang.org/stable/std/primitive.slice.html#method.copy_from_slice here instead and avoid unsafe code.